nengo_gui/sim.py

```diff
@@ -0,0 +1,435 @@
+        self.rebuild = False   # should the model be rebuilt
+
+        self.undo_stack = []
+        self.redo_stack = []
+
+        self.config_save_period = 2.0  # minimum time between saves
+
+        self.lock = threading.Lock()
+
+        # use the default filename if none is given
+        if filename is None:
+            self.filename = sim_server.filename
+        else:
+            self.filename = os.path.relpath(filename)
+
+        # determine the .cfg filename
+        if sim_server.filename_cfg is None:
+            self.filename_cfg = self.filename + '.cfg'
+        else:
+            self.filename_cfg = sim_server.filename_cfg
+
+        if reset_cfg:
+            self.clear_config()
+
+        self.load()
+
+        self.net_graph = self.get_net_graph()
+
+        # build and run the model in a separate thread
+        t = threading.Thread(target=self.runner)
+        t.daemon = True
+        t.start()
+
+    @property
+    def sim(self):
+        return self._sim
+
+    @sim.setter
+    def sim(self, value):
+        if hasattr(self._sim, 'close'):
+            self._sim.close()
+        self._sim = value
+
+    def get_net_graph(self):
+        for c in self.components:
+            if isinstance(c, nengo_gui.components.NetGraph):
+                return c
+        return None
+
+    def clear_config(self):
+        if os.path.isfile(self.filename_cfg):
+            os.remove(self.filename_cfg)
+
+    def load(self):
+        """Load the model and initialize everything"""
+        if self.filename == self.sim_server.filename:
+            # if we're on the default filenaem, just load it from the SimServer
+            self.model = self.sim_server.model
+            if self.sim_server.locals is None:
+                self.locals = None
+            else:
+                self.locals = self.sim_server.locals.copy()
+        else:
+            self.model = None
+            self.locals = None
+
+        # if we still don't have a locals dictionary, then run the script
+        if self.locals is None:
+            try:
+                with open(self.filename) as f:
+                    code = f.read()
+            except IOError:
+                code = ''
+
+            self.execute(code)
+
+        if self.model is None:
+            self.model = nengo.Network()
+            self.locals['model'] = self.model
+
+        # figure out good names for objects
+        self.name_finder = nengo_gui.NameFinder(self.locals, self.model)
+        self.default_labels = self.name_finder.known_name
+
+        # load the .cfg file
+        self.config = self.load_config()
+        self.config_save_needed = False
+        self.config_save_time = None   # time of last config file save
+
+        self.uid_prefix_counter = {}
+
+        self.create_components()
+
+    def create_components(self):
+        """Generate the actual Components from the Templates"""
+        self.components = []
```

Comment:
There are a number of attributes (this is just the first one) in this class which are not defined within the `__init__` function which would be preferable (especially if there are obvious default values like the empty list). That would also help with auto-completion in some editors.